### PR TITLE
[frieren] Round-7: Lion uncompiled SOTA + 1-epoch LR warmup

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

**Lion uncompiled SOTA + 1-epoch LR warmup.** Single-variable test vs PR #50 (current SOTA 11.208).

PR #50 (vanilla Lion uncompiled, no warmup) achieves test_abupt 11.208. Lion uses `sign()` updates with constant per-coordinate step magnitude bounded by `lr`. A 1-epoch warmup linearly ramps from lr=0 → 5e-5, giving:
1. Reduced early-epoch step-size pressure when gradients are large/noisy
2. Smoother convergence entry before the model settles into its attractor basin
3. Untested on Lion uncompiled — could improve final val by 0.1-0.5 ppt

No schedule after warmup (no cosine). Just warmup, then constant lr=5e-5. Orthogonal to nezuko #93 (Lion + cosine T_max=24) which tests the schedule tail.

## Instructions

Only `--lr-warmup-epochs 1` is added vs PR #50. Do NOT change anything else.

```bash
torchrun --standalone --nproc_per_node=8 train.py \
  --agent frieren \
  --optimizer lion \
  --lr 5e-5 --weight-decay 5e-4 \
  --no-compile-model \
  --batch-size 4 \
  --validation-every 1 \
  --train-surface-points 65536 \
  --eval-surface-points 65536 \
  --train-volume-points 65536 \
  --eval-volume-points 65536 \
  --model-layers 4 \
  --model-hidden-dim 512 \
  --model-heads 8 \
  --model-slices 128 \
  --ema-decay 0.9995 \
  --lr-warmup-epochs 1 \
  --wandb-group tay-round7-lion-warmup
```

Train for full 270min budget. Lion uncompiled gets ~9 epochs. Submit when training + test eval completes.

## Baseline (current SOTA — PR #50 nezuko, W&B run `g2n4fyta`)

| Metric | PR #50 SOTA | AB-UPT target |
|---|---:|---:|
| `test_primary/abupt_axis_mean_rel_l2_pct` | **11.208** | — |
| `surface_pressure` | 6.193 | 3.82 |
| `wall_shear` | 11.199 | 7.29 |
| `volume_pressure` | 12.726 | 6.08 |
| `tau_x` | 9.512 | 5.35 |
| `tau_y` | 13.592 | 3.65 |
| `tau_z` | 14.017 | 3.63 |

PR #50 config: Lion lr=5e-5 wd=5e-4, 4L/512d/8h/128slices, no compile, EMA 0.9995, no schedule, no warmup. Best val 10.08 at ep9, 287min runtime.

**Decision criteria:**
- MERGE if test_abupt < 11.208
- REQUEST CHANGES if test 11.208–11.50 (try warmup_epochs=2 or warmup + cosine)
- CLOSE if test > 11.50 (warmup hurts)

### CODE REVIEW

**Comment by morganmcg1** [comment]
STUDENT frieren:

Pre-run notes — deviations from the literal PR command, made to honor the meta-instruction *"Only `--lr-warmup-epochs 1` is added vs PR #50. Do NOT change anything else."*:

1. **Adding `--volume-loss-weight 2.0`** — PR #50 used this explicitly (per BASELINE.md). The PR #109 command drops it (default would be 1.0), which would change two variables instead of one. Restoring 2.0 to keep this a true single-delta test.
2. **Adding `--lion-beta1 0.9 --lion-beta2 0.99`** — defaults already match PR #50, but making explicit for safety.
3. **Adding logging flags** `--gradient-log-every 100 --weight-log-every 100 --no-log-gradient-histograms` — match PR #50's logging cadence.

**Heads-up on schedule:** PR description says "No schedule after warmup (no cosine). Just warmup, then constant lr=5e-5." — but `build_lr_scheduler` in trainer_runtime.py always wraps in `CosineAnnealingLR` after warmup (or as the only schedule when warmup=0). So both PR #50 and this run have cosine annealing 5e-5 → 1e-6 as the default; the literal single delta is just "prepend 1-epoch linear warmup (start_factor=0.05) before the existing cosine." If the advisor truly wants warmup-then-constant-LR, that requires a code change and `--lr-min 5e-5`. Will run with the existing scheduler behavior unless instructed otherwise.

**Budget:** `SENPAI_TIMEOUT_MINUTES=360` (PR mentions 270min — fleet appears to be on 6h budget now). Will report runtime + epochs achieved.

Launching run now.

**Comment by morganmcg1** [comment]
STUDENT frieren:

## Results

**Outcome: Negative — warmup hurt vs PR #50 SOTA.**

### Test metrics (vs PR #50 SOTA 11.208)

| Metric | PR #50 SOTA | PR #109 (this) | Δ |
|---|---:|---:|---:|
| `test_primary/abupt_axis_mean_rel_l2_pct` | **11.208** | **11.596** | **+0.388** |
| `surface_pressure` | 6.193 | 6.540 | +0.347 |
| `wall_shear` | 11.199 | 11.647 | +0.448 |
| `volume_pressure` | 12.726 | 12.857 | +0.131 |
| `tau_x` | 9.512 | 9.910 | +0.398 |
| `tau_y` | 13.592 | 14.100 | +0.508 |
| `tau_z` | 14.017 | 14.572 | +0.555 |

Worse on every axis. Lands in the **CLOSE** decision band (test 11.596 > 11.50).

### Validation metrics

| Metric | PR #50 SOTA | PR #109 (this) | Δ |
|---|---:|---:|---:|
| `best_val_primary/abupt_axis_mean_rel_l2_pct` | 10.080 | 10.415 | +0.335 |
| best epoch | 9 | 9 | — |
| `val_primary/surface_pressure_rel_l2_pct` | — | 6.757 | — |
| `val_primary/volume_pressure_rel_l2_pct` | — | 6.031 | — |
| `val_primary/wall_shear_rel_l2_pct` | — | 11.741 | — |

Per-epoch val_primary abupt trajectory:
- ep1: 84.62, ep2: 67.37, ep3: 39.62, ep4: 21.11, ep5: 15.43,
- ep6: 12.90, ep7: 11.55, ep8: 10.77, **ep9: 10.42 (best)**

Final val slope ~−0.22 ppt/1k steps — model still slowly improving, did not plateau.

### Run details

- **Command:**
  ```bash
  torchrun --standalone --nproc_per_node=8 train.py \
    --agent frieren --optimizer lion --lion-beta1 0.9 --lion-beta2 0.99 \
    --lr 5e-5 --weight-decay 5e-4 --no-compile-model \
    --volume-loss-weight 2.0 --batch-size 4 --validation-every 1 \
    --train-surface-points 65536 --eval-surface-points 65536 \
    --train-volume-points 65536 --eval-volume-points 65536 \
    --model-layers 4 --model-hidden-dim 512 --model-heads 8 --model-slices 128 \
    --ema-decay 0.9995 --gradient-log-every 100 --weight-log-every 100 \
    --no-log-gradient-histograms --lr-warmup-epochs 1 \
    --wandb-group tay-round7-lion-warmup --wandb-name frieren/lion-warmup1
  ```
- **Runtime:** 270.8 min (matched PR #50's 287 min budget; got 9 epochs, same as PR #50)
- **Peak GPU memory:** ~74.2 GB/GPU (8× H100, max from W&B `system.gpu.*.memoryAllocatedBytes`)
- **W&B run ID:** `kfy5sykb` (rank 0 lead) — group `tay-round7-lion-warmup`
  - https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml-ddp8/runs/kfy5sykb

### What happened

The 1-epoch linear warmup (start_factor=0.05, ramp 2.5e-6 → 5e-5) **delayed convergence by ~1 epoch worth of effective updates** without any compensating gain in final-epoch behavior. Both runs converged for 9 epochs, but the warmup run reaches the same trajectory shape ~1 epoch later — the model never caught up before the budget cut.

**Why warmup didn't help Lion here:**
1. **Lion already self-bounds early-step magnitude.** Lion's `sign(momentum)` update has constant per-coordinate step bounded by `lr` regardless of gradient scale. So the standard warmup motivation — taming early-step blow-ups when gradients are large/noisy — is partially neutralized. Warmup mainly costs us optimization budget at small effective LR.
2. **No instability to fix.** PR #50's grad-norm trace shows no early-epoch divergence or NaN events at lr=5e-5. There was nothing for warmup to stabilize.
3. **Schedule note (already flagged pre-run).** As I noted before launch, `build_lr_scheduler` always wraps in CosineAnnealingLR after warmup, so the literal `--lr-warmup-epochs 1` adds warmup *before* the existing cosine schedule. The PR description's "no schedule after warmup" framing wasn't realized in code; both this run and PR #50 have cosine-from-5e-5-to-1e-6. The single delta is genuinely just the prepended 1-epoch warmup.

The +0.39 ppt regression is consistent with "lost ~1 epoch of training to a slow ramp" given the ~−0.4 ppt/epoch slope observed at ep9.

### Suggested follow-ups

- **CLOSE per decision criteria** (test > 11.50). Warmup is not the right lever for Lion at this scale/budget.
- If the advisor still wants to test warmup behavior, a fairer test would be **shorter warmup (200–500 steps, not 1 epoch)** or **warmup + extended budget** to compensate for the lost epoch. But those are different hypotheses.
- Lion uncompiled stays the SOTA at 11.208 (PR #50). Better leverage points likely live elsewhere (architecture depth, loss weighting, EMA decay, data augmentation), not in early-LR scheduling.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #109 because ADVISOR: Closing as confirmed regression. Test 11.596 vs SOTA 11.208 = +3.5%, regresses on every axis. **Warmup is closed-door on Lion uncompiled stack** — 1-epoch LR warmup hurts more than it helps in our 9-epoch budget. Mechanistically: Lion's sign-based updates are already low-magnitude relative to AdamW, so warmup further reduces effective learning during a phase where we can't afford slow descent. Clean execution by frieren — single-delta test, all deviations correctly noted. Reassigning to next hypothesis..
